[addons] add all needed addon window interface calls

AlwinEsch opened the pull request:
## Description
<!--- Describe your change in detail -->
Next step for binary addon interface rework.

This is nearly the last change on addon Window class. There becomes
all parts added to have needed functionality, also is it with supported
parts now nearly equal with a Python addon window interface.

As last changes on window part comes on next request only the support
for the various control types.

<!--- Provide a general summary of your change in the Title above -->


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

ksooo commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+  if (!pWindow)
+    return -1;
+
+
```
Could you please remove one of the two empty lines here and everywhere else in this source file. Seen thin in several places from here.

ksooo commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+
+  std::string lowerKey = key;
+  StringUtils::ToLower(lowerKey);
+  Interface_GUIGeneral::lock();
```
To follow your own style you should insert an empty line before this call (after the tolower call). :-)

ksooo commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+    return false;
+  }
+
+  CGUIWindow *pWindow = dynamic_cast<CGUIWindow*>(g_windowManager.GetWindow(pAddonWindow->m_windowId));
```
If pAddonWindow is of type <code>CGUIAddonWindow</code> which is guaranteed if you come across this code line, it is save that pAddonWindow is a <code>CGUIWindow</code>, because <code>CGUIAddonWindow</code> is derived from <code>CGUIMediaWindow</code> which is derived from <code>CGUIWindow</code>, right? You don't need pWindow here and everywhere in the code you introduced in this PR. You can call directly on pAddonWindow. ;-)

AlwinEsch replied:
The path is a remnant of the beginning

https://github.com/xbmc/xbmc/blob/4e9cb43928abbddd5d683cb5d9d66c72b3bbf88e/xbmc/addons/AddonCallbacksGUI.cpp#L388_L391

I think is to confirm Window registered with `g_windowManager.Add(window);`.

But from my side it can also be removed.

ksooo commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+    return;
+  }
+
+  pAddonWindow->MarkDirtyRegion();
```
Again, why no lock/unlock before/after this call?

ksooo commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h:
```diff
@@ -175,9 +176,39 @@ typedef struct AddonToKodiFuncTable_kodi_gui_window
+  void (*set_container_content)(void* kodiBase, void* handle, const char* value);
+  int (*get_current_container_id)(void* kodiBase, void* handle);
+
+  /* Various calls */
```
minor nitpicking: these are not calls, but functions. Seen multiple times in this source file.

AlwinEsch replied:
Upps, had forgotten the place

ksooo reviewed (changes requested):
As  usual I have some small questions and suggestions.

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+  if (control_id == -1)
+    CLog::Log(LOGERROR, "Interface_GUIWindow - %s: %s - No control in this window has focus", __FUNCTION__, addon->Name().c_str());
+
+  return control_id;
```
Upss, was a dead corpse.

ksooo replied:
Okay.

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+
+  Interface_GUIGeneral::lock();
+  CGUIMessage msg = CGUIMessage(GUI_MSG_SETFOCUS, pAddonWindow->m_windowId, control_id);
+  pWindow->OnMessage(msg);
```
Have testet, compiler does not accept to have in OnMessage, I change it to 
```cpp
  CGUIMessage msg(GUI_MSG_SETFOCUS, pAddonWindow->m_windowId, control_id);
  pWindow->OnMessage(msg);
```
to match the others.

AlwinEsch committed:
[addons] add all needed addon window interface calls

Next step for binary addon interface rework.

This is nearly the last change on addon Window class. There becomes
all parts added to have needed functionality, also is it with supported
parts now nearly equal with a Python addon window interface.

As last changes on window part comes on next request only the support
for the various control types.

AlwinEsch commented:
Is updated

AlwinEsch commented:
jenkins build this please

AchimTuran commented:
@AlwinEsch 
I reviewed your PR. If you have further questions or suggestions please ping on Github, Slack... 

Only minor issues from my side and it is up to you to accept or reject them.

Again thank you for your work, I'm really looking forward to use it in my upcoming AudioDSP binary add-on dialogs.

Keep up the good work.

ksooo commented:
jenkins build and merge

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -316,11 +625,50 @@ void Interface_GUIWindow::add_list_item(void* kodiBase, void* handle, void* item
 
   Interface_GUIGeneral::lock();
-  pAddonWindow->AddItem(pItem, item_position);
+  pAddonWindow->AddItem(pItem, list_position);
```
The use as "list_position" is to match the for "C" standardized string format. The other is normally frowned upon for "C".

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h:
```diff
@@ -166,6 +166,232 @@ namespace gui
+      if (ret != nullptr)
+      {
+        if (std::strlen(ret))
+          label = ret;
```
Is a bit to prevent the C++ class construction and copy from the not present string, not complete sure what becomes done in compiled code, but thought why let C++ work if C nothing have.

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h:
```diff
@@ -175,9 +176,39 @@ typedef struct AddonToKodiFuncTable_kodi_gui_window
   void (*clear_item_list)(void* kodiBase, void* handle);
-  void (*add_list_item)(void* kodiBase, void* handle, void* item, int item_position);
-  void* (*get_list_item)(void* kodiBase, void* handle, int listPos);
+  void (*add_list_item)(void* kodiBase, void* handle, void* item, int list_position);
```
Has leaved only "kodiBase" in camel case to see for me that is C++ on Kodi related.

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -280,7 +315,281 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+bool Interface_GUIWindow::set_focus_id(void* kodiBase, void* handle, int control_id)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUIAddonWindow* pAddonWindow = static_cast<CGUIAddonWindow*>(handle);
```
@ksooo what you thing about `reinterpret_cast`?
For me, this is a final solution if nothing else is possible.

ksooo replied:
Whenever you cast a void to something else, imo, reinterpret_cast is the best choice.

AlwinEsch commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -280,7 +315,281 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
+  StringUtils::ToLower(lowerKey);
+
+  Interface_GUIGeneral::lock();
+  pAddonWindow->SetProperty(lowerKey, "");
```
This reset the property on given key value.

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h:
```diff
@@ -166,6 +166,232 @@ namespace gui
+    ///        optional arguments.\n
+    ///        Once you use a keyword, all following arguments require the keyword.
+    ///
+    void SetProperty(const std::string& key, const std::string& value)
```
Thanks! @ksooo has brought me to them.

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h:
```diff
@@ -227,6 +481,135 @@ namespace gui
+    /// @param[in] listPos              Position to use
+    ///
+    ///
+    void SetCurrentListPosition(int listPos)
```
I look, maybe yes.

AlwinEsch commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h:
```diff
@@ -227,6 +481,135 @@ namespace gui
+    /// @brief To inform Kodi that it need to render region new.
+    ///
+    ///
+    void MarkDirtyRegion()
```
No Problem :)
With the documentation is currently added what I done some time ago. Currently only present but not direct adjusted for doxygen. Because of the size comes the optimization on separate requests.

ksooo commented on xbmc/addons/interfaces/GUI/Window.cpp:
```diff
@@ -348,6 +696,130 @@ void* Interface_GUIWindow::get_list_item(void* kodiBase, void* handle, int listP
+  }
+
+  Interface_GUIGeneral::lock();
+  pAddonWindow->SetCurrentListPosition(list_position);
```
I see it vica versa - for C functions, I would use underline names, for C++ camel case. We're not consistent here in any way.

But may I suggest to do this kind of cleanup in a separate PR, once all stuff from Alwin is in? I think we agreed on this already.
